### PR TITLE
[Fix] Fix missing `state_dict._metadata` when saving and loading checkpoints.

### PR DESCRIPTION
**Pull request by mzr1996**
## Motivation

PyTorch provides `_version` in `nn.Module` to support backward compatibility. However, because of some modifications in saving & loading in mmcv, when overriding the `_load_from_state_dict` in a module, the `local_metadata` is empty dict now.

## Modification

Modify some code in saving and loading to keep `state_dict._metadata`.

## BC-breaking (Optional)

No, it should be beneficial for backward compatibility, because downstream repos, like `mmdet`, are using these metadata to support old version checkpoints, and this function is not available now.

## Use cases (Optional)

```python
>>> import torch
>>> import torch.nn as nn
>>> from mmcv.runner import save_checkpoint
>>> 
>>> class ModelV2(nn.Module):
...     _version = 2
... 
...     def __init__(self):
...         super().__init__()
...         self.conv = nn.Conv2d(3, 3, 1)
...
>>> save_checkpoint(ModelV2(), 'test.pth')
>>> model = torch.load('test.pth')
>>> print(model['state_dict']._metadata)
OrderedDict([('', {'version': 2}), ('conv', {'version': 1})])
```

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: tests/test_runner/test_checkpoint.py [hunk 1]
```diff
@@ -221,6 +221,77 @@ def __init__(self):
...
+            nn.init.normal_(self.conv1.weight)
+
+        def _load_from_state_dict(self, state_dict, prefix, local_metadata,
+                                  *args, **kwargs):
+            """load checkpoints."""
+
+            # Names of some parameters in has been changed.
+            version = local_metadata.get('version', None)
+            if version is None or version < 2:
+                state_dict_keys = list(state_dict.keys())
```
Comment (zhouzaida) [inline tests/test_runner/test_checkpoint.py:258]:
```python
                state_dict_keys = list(state_dict.keys())
                for k in state_dict_keys:
                    convert_map = {'conv1': 'conv0', 'conv2': 'conv1'}
                    for ori_pattern, convert_pattern in convert_map.items():
                        if k.startswith(prefix + ori_pattern):
                            convert_key = k.replace(ori_pattern,
                                                    convert_pattern)
                            state_dict[convert_key] = state_dict[k]
                            del state_dict[k]
```

-> 

```python
                convert_map = {'conv1': 'conv0', 'conv2': 'conv1'}
                for k in state_dict.keys():
                    for ori_str, new_str in convert_map.items():
                        if k.startswith(prefix + ori_str):
                            new_key = k.replace(ori_str, new_str)
                            state_dict[new_key] = state_dict[k]
                            del state_dict[k]
```

Comment (mzr1996) [reply tests/test_runner/test_checkpoint.py:258]: I think we cannot use `state_dict.keys()` in iteration.
For example:
```python
>>> d = dict(a=1, b=2, c=3)
>>> for k in d.keys():
...     print(k)
...     del d[k]
a
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-2-555253fa01ee> in <module>
----> 1 for k in d.keys():
      2     print(k)
      3     del d[k]

RuntimeError: dictionary changed size during iteration
```

Comment (mzr1996) [reply tests/test_runner/test_checkpoint.py:258]: The variable names change requests have been done.

Comment (zhouzaida) [reply tests/test_runner/test_checkpoint.py:258]: got it

File: tests/test_runner/test_checkpoint.py [hunk 2]
```diff
@@ -221,6 +221,76 @@ def __init__(self):
...
+                convert_map = {'conv1': 'conv0', 'conv2': 'conv1'}
+                for k in state_dict_keys:
+                    for ori_str, new_str in convert_map.items():
+                        if k.startswith(prefix + ori_str):
+                            new_key = k.replace(ori_str, new_str)
+                            state_dict[new_key] = state_dict[k]
+                            del state_dict[k]
```
Comment (zhouzaida) [inline tests/test_runner/test_checkpoint.py:265]:
```suggestion
                new_state_dict = {}
                key_mappings = {'conv1': 'conv0', 'conv2': 'conv1'}
                for k in state_dict.keys():
                    for ori_str, new_str in key_mappings.items():
                        if k.startswith(prefix + ori_str):
                            new_key = k.replace(ori_str, new_str)
                            new_state_dict[new_key] = state_dict[k]
                            break
                    else:
                        new_state_dict[k] = state_dict[k]
```

Comment (mzr1996) [reply tests/test_runner/test_checkpoint.py:265]: Perhaps we cannot apply it.
https://github.com/open-mmlab/mmcv/blob/d3b057240829a0662125d5e83e002b33cd3fa874/mmcv/runner/checkpoint.py#L66-L79
When loading the checkpoint, we recursively call `_load_from_state_dict` from each module. If we don't inplace modify the state_dict, this modification will be ineffective.

Comment (zhouzaida) [reply tests/test_runner/test_checkpoint.py:265]: got it

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**